File: scripts/upload_nsm_churches.js
const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');
const serviceAccount = require('../pastor-report-e4c52-firebase-adminsdk-fbsvc-a7f1335a9b.json');

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  projectId: 'pastor-report-e4c52'
});

const db = admin.firestore();

// Read the NSM Churches data
const nsmChurchesPath = path.join(__dirname, '../assets/NSM_Churches_Updated.json');
const nsmData = JSON.parse(fs.readFileSync(nsmChurchesPath, 'utf8'));

const MISSION_ID = 'north_sabah_mission';
const MISSION_NAME = 'North Sabah Mission';

async function uploadNSMChurches() {
  console.log('🚀 Uploading North Sabah Mission Churches...\n');

  try {
    // Step 1: Ensure mission exists
    console.log('📋 Step 1: Checking/Creating Mission...');
    const missionRef = db.collection('missions').doc(MISSION_ID);
    const missionDoc = await missionRef.get();

    if (!missionDoc.exists) {
      await missionRef.set({
        id: MISSION_ID,
        name: MISSION_NAME,
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
      });
      console.log(`  ✓ Created mission: ${MISSION_NAME}`);
    } else {
      console.log(`  ✓ Mission exists: ${MISSION_NAME}`);
    }

    let totalRegions = 0;
    let totalDistricts = 0;
    let totalChurches = 0;
    let totalCompanies = 0;
    let totalGroups = 0;

    // Step 2: Process each region
    console.log('\n📋 Step 2: Processing Regions and Districts...\n');

    for (const [regionNum, regionData] of Object.entries(nsmData.regions)) {
      const regionId = `nsm_region_${regionNum}`;
      const regionName = regionData.name;

      // Create/Update region
      const regionRef = db.collection('regions').doc(regionId);
      await regionRef.set({
        id: regionId,
        name: regionName,
        missionId: MISSION_ID,
        createdAt: admin.firestore.FieldValue.serverTimestamp(),
      }, { merge: true });

      console.log(`✓ Region: ${regionName}`);
      totalRegions++;

      // Process pastoral districts
      for (const [districtName, districtData] of Object.entries(regionData.pastoral_districts)) {
        const districtId = `nsm_district_${regionNum}_${districtName.toLowerCase().replace(/[^a-z0-9]+/g, '_')}`;

        // Create/Update district
        const districtRef = db.collection('districts').doc(districtId);
        await districtRef.set({
          id: districtId,
          name: districtName,
          regionId: regionId,
          missionId: MISSION_ID,
          pastor: districtData.pastor || '',
          phone: districtData.phone || '',
          createdAt: admin.firestore.FieldValue.serverTimestamp(),
        }, { merge: true });

        console.log(`  ✓ District: ${districtName} (Pastor: ${districtData.pastor || 'TBD'})`);
        totalDistricts++;

        // Process organized churches
        if (districtData.organized_churches && districtData.organized_churches.length > 0) {
          for (const church of districtData.organized_churches) {
            const churchId = `nsm_church_${districtId}_${church.name.toLowerCase().replace(/[^a-z0-9]+/g, '_')}`;

            await db.collection('churches').doc(churchId).set({
              id: churchId,
              churchName: church.name,
              districtId: districtId,
              regionId: regionId,
              missionId: MISSION_ID,
              status: 'church',
              userId: 'system',
              elderName: '',
              elderEmail: '',
              elderPhone: '',
              memberCount: 0,
              dateOrganized: church.doc || null,
              createdAt: admin.firestore.FieldValue.serverTimestamp(),
            }, { merge: true });

            console.log(`    ⛪ Church: ${church.name}`);
            totalChurches++;
          }
        }

        // Process companies
        if (districtData.companies && districtData.companies.length > 0) {
          for (const company of districtData.companies) {
            const companyId = `nsm_company_${districtId}_${company.name.toLowerCase().replace(/[^a-z0-9]+/g, '_')}`;

            await db.collection('churches').doc(companyId).set({
              id: companyId,
              churchName: company.name,
              districtId: districtId,
              regionId: regionId,
              missionId: MISSION_ID,
              status: 'company',
              userId: 'system',
              elderName: '',
              elderEmail: '',
              elderPhone: '',
              memberCount: 0,
              createdAt: admin.firestore.FieldValue.serverTimestamp(),
            }, { merge: true });

            console.log(`    🏢 Company: ${company.name}`);
            totalCompanies++;
          }
        }

        // Process groups
        if (districtData.groups && districtData.groups.length > 0) {
          for (const group of districtData.groups) {
            const groupId = `nsm_group_${districtId}_${group.name.toLowerCase().replace(/[^a-z0-9]+/g, '_')}`;

            await db.collection('churches').doc(groupId).set({
              id: groupId,
              churchName: group.name,
              districtId: districtId,
              regionId: regionId,
              missionId: MISSION_ID,
              status: 'group',
              userId: 'system',
              elderName: '',
              elderEmail: '',
              elderPhone: '',
              memberCount: 0,
              createdAt: admin.firestore.FieldValue.serverTimestamp(),
            }, { merge: true });

            console.log(`    👥 Group: ${group.name}`);
            totalGroups++;
          }
        }
      }

      console.log('');
    }

    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('📊 UPLOAD SUMMARY');
    console.log('='.repeat(60));
    console.log(`✅ Mission: ${MISSION_NAME}`);
    console.log(`✅ Regions: ${totalRegions}`);
    console.log(`✅ Districts: ${totalDistricts}`);
    console.log(`⛪ Organized Churches: ${totalChurches}`);
    console.log(`🏢 Companies: ${totalCompanies}`);
    console.log(`👥 Groups: ${totalGroups}`);
    console.log(`📈 Total Congregations: ${totalChurches + totalCompanies + totalGroups}`);
    console.log('='.repeat(60));

    console.log('\n🎉 Upload complete! Check your Firebase console to verify the data.');

    process.exit(0);
  } catch (error) {
    console.error('\n❌ Error:', error);
    process.exit(1);
  }
}

uploadNSMChurches();

File: scripts/upload_nsm_churches.dart
import 'dart:convert';
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';

const String missionId = 'north_sabah_mission';
const String missionName = 'North Sabah Mission';

Future<void> main() async {
  print('🚀 Uploading North Sabah Mission Churches...\n');

  try {
    // Initialize Firebase
    await Firebase.initializeApp();
    final db = FirebaseFirestore.instance;

    // Read the NSM Churches data
    final file = File('assets/NSM_Churches_Updated.json');
    final jsonString = await file.readAsString();
    final nsmData = json.decode(jsonString) as Map<String, dynamic>;

    // Step 1: Ensure mission exists
    print('📋 Step 1: Checking/Creating Mission...');
    final missionRef = db.collection('missions').doc(missionId);
    final missionDoc = await missionRef.get();

    if (!missionDoc.exists) {
      await missionRef.set({
        'id': missionId,
        'name': missionName,
        'createdAt': FieldValue.serverTimestamp(),
      });
      print('  ✓ Created mission: $missionName');
    } else {
      print('  ✓ Mission exists: $missionName');
    }

    int totalRegions = 0;
    int totalDistricts = 0;
    int totalChurches = 0;
    int totalCompanies = 0;
    int totalGroups = 0;

    // Step 2: Process each region
    print('\n📋 Step 2: Processing Regions and Districts...\n');

    final regions = nsmData['regions'] as Map<String, dynamic>;

    for (var regionEntry in regions.entries) {
      final regionNum = regionEntry.key;
      final regionData = regionEntry.value as Map<String, dynamic>;
      final regionId = 'nsm_region_$regionNum';
      final regionName = regionData['name'] as String;

      // Create/Update region
      await db.collection('regions').doc(regionId).set({
        'id': regionId,
        'name': regionName,
        'missionId': missionId,
        'createdAt': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));

      print('✓ Region: $regionName');
      totalRegions++;

      // Process pastoral districts
      final pastoralDistricts =
          regionData['pastoral_districts'] as Map<String, dynamic>;

      for (var districtEntry in pastoralDistricts.entries) {
        final districtName = districtEntry.key;
        final districtData = districtEntry.value as Map<String, dynamic>;
        final districtId =
            'nsm_district_${regionNum}_${districtName.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]+'), '_')}';

        // Create/Update district
        await db.collection('districts').doc(districtId).set({
          'id': districtId,
          'name': districtName,
          'regionId': regionId,
          'missionId': missionId,
          'pastor': districtData['pastor'] ?? '',
          'phone': districtData['phone'] ?? '',
          'createdAt': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));

        print(
            '  ✓ District: $districtName (Pastor: ${districtData['pastor'] ?? 'TBD'})');
        totalDistricts++;

        // Process organized churches
        final organizedChurches =
            districtData['organized_churches'] as List<dynamic>? ?? [];
        for (var church in organizedChurches) {
          final churchMap = church as Map<String, dynamic>;
          final churchName = churchMap['name'] as String;
          final churchId =
              'nsm_church_${districtId}_${churchName.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]+'), '_')}';

          await db.collection('churches').doc(churchId).set({
            'id': churchId,
            'churchName': churchName,
            'districtId': districtId,
            'regionId': regionId,
            'missionId': missionId,
            'status': 'church',
            'userId': 'system',
            'elderName': '',
            'elderEmail': '',
            'elderPhone': '',
            'memberCount': 0,
            'dateOrganized': churchMap['doc'],
            'createdAt': FieldValue.serverTimestamp(),
          }, SetOptions(merge: true));

          print('    ⛪ Church: $churchName');
          totalChurches++;
        }

        // Process companies
        final companies = districtData['companies'] as List<dynamic>? ?? [];
        for (var company in companies) {
          final companyMap = company as Map<String, dynamic>;
          final companyName = companyMap['name'] as String;
          final companyId =
              'nsm_company_${districtId}_${companyName.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]+'), '_')}';

          await db.collection('churches').doc(companyId).set({
            'id': companyId,
            'churchName': companyName,
            'districtId': districtId,
            'regionId': regionId,
            'missionId': missionId,
            'status': 'company',
            'userId': 'system',
            'elderName': '',
            'elderEmail': '',
            'elderPhone': '',
            'memberCount': 0,
            'createdAt': FieldValue.serverTimestamp(),
          }, SetOptions(merge: true));

          print('    🏢 Company: $companyName');
          totalCompanies++;
        }

        // Process groups
        final groups = districtData['groups'] as List<dynamic>? ?? [];
        for (var group in groups) {
          final groupMap = group as Map<String, dynamic>;
          final groupName = groupMap['name'] as String;
          final groupId =
              'nsm_group_${districtId}_${groupName.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]+'), '_')}';

          await db.collection('churches').doc(groupId).set({
            'id': groupId,
            'churchName': groupName,
            'districtId': districtId,
            'regionId': regionId,
            'missionId': missionId,
            'status': 'group',
            'userId': 'system',
            'elderName': '',
            'elderEmail': '',
            'elderPhone': '',
            'memberCount': 0,
            'createdAt': FieldValue.serverTimestamp(),
          }, SetOptions(merge: true));

          print('    👥 Group: $groupName');
          totalGroups++;
        }
      }

      print('');
    }

    // Summary
    print('\n${'=' * 60}');
    print('📊 UPLOAD SUMMARY');
    print('=' * 60);
    print('✅ Mission: $missionName');
    print('✅ Regions: $totalRegions');
    print('✅ Districts: $totalDistricts');
    print('⛪ Organized Churches: $totalChurches');
    print('🏢 Companies: $totalCompanies');
    print('👥 Groups: $totalGroups');
    print('📈 Total Congregations: ${totalChurches + totalCompanies + totalGroups}');
    print('=' * 60);

    print(
        '\n🎉 Upload complete! Check your Firebase console to verify the data.');
  } catch (error) {
    print('\n❌ Error: $error');
    exit(1);
  }

  exit(0);
}
